### DISCUSSION TITLE
Enums as type constraints?

### DISCUSSION THREAD

**Discussion by marwalsch** [dotnet/csharplang#5522] [Q&A] [answered]
There are plenty of issues on Stack Overflow regarding how `System.Enum` can/could be used as a type constraint, but I have found no discussion/solution on passing a certain enum as a type constraint. I stumbled across this by trying to implement the following:

```
    public class HttpStatusException<T> : Exception where T : HttpStatusCode
    { 
        //...
    }
```

This does currently not work, but could allow for constructs like the following:

```
try
{
   //...  
}catch(HttpStatusException<HttpStatusCode.Forbidden> ex)
{
   //...
}
```

I couldn't think about a lot of other use cases for this, but is this something feasible/desirable to implement?

**Comment by FiniteReality** [comment, marked as answer]
It would require runtime changes, which basically puts this at -1000 points, rather than the -100 points other changes would require. Which is unfortunate, because the more usable form of this - integer constants - would have lots of potential uses such as matrices or vectors which are implemented in a generic way regardless of the number of elements.

**Comment by marwalsch** [comment]
For those stumbling across this upon their search for a smooth solution for my initial problem, there is a [when condition for catch blocks ](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/when)starting from C# 6.
